<a href="https://colab.research.google.com/github/jackieuxui/telecom-plan-recommendation-ml/blob/main/Intro_to_Machine_Learning_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Intro to Machine Learning: Project
Congratulations! You’ve completed another training platform course. Now is the perfect time to test your skills and solve a new machine learning problem. For this project, you will be on your own.

When you finish, send your work to the project instructor. You will receive feedback within 24 hours. After that, you will make any necessary changes to you work and send it for a second review.

Usually this process has to be repeated several times until you get the green light from the instructor and all the corrections are approved. That’s all part of the job.

#Project description
Mobile carrier Megaline has found out that many of their subscribers use legacy plans. They want to develop a model that would analyze subscribers' behavior and recommend one of Megaline's newer plans: Smart or Ultra.

You have access to behavior data about subscribers who have already switched to the new plans (from the project for the Statistical Data Analysis course). For this classification task, you need to develop a model that will pick the right plan. Since you’ve already performed the data preprocessing step, you can move straight to creating the model.  

Develop a model with the highest possible accuracy. In this project, the threshold for accuracy is 0.75. Check the accuracy using the test dataset.  

### Project instructions
Open and look through the data file. Path to the file:datasets/users_behavior.csv Download dataset
Split the source data into a training set, a validation set, and a test set.
Investigate the quality of different models by changing hyperparameters. Briefly describe the findings of the study.
Check the quality of the model using the test set.
Additional task: sanity check the model. This data is more complex than what you’re used to working with, so it's not an easy task. We'll take a closer look at it later.
### Data description
Every observation in the dataset contains monthly behavior information about one user. The information given is as follows:

* сalls — number of calls,
* minutes — total call duration in minutes,
* messages — number of text messages,
* mb_used — Internet traffic used in MB,
* is_ultra — plan for the current month (Ultra - 1, Smart - 0).

# Project evaluation
We’ve put together the evaluation criteria for the project. Read this carefully before moving on to the task.

Here’s what the instructors will look at when reviewing your project:

* How did you look into data after downloading?
* Have you correctly split the data into train, validation, and test sets?
* How have you chosen the sets' sizes?
* Did you evaluate the quality of the models correctly?
* What models and hyperparameters did you use?
* What are your findings?
* Did you test the models correctly?
* What is your accuracy score?
* Have you stuck to the project structure and kept the code neat?

# Import libraries

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [2]:
# Upload document
from google.colab import files
uploaded = files.upload()
# Check file directory
import os
print(os.listdir())

Saving users_behavior.csv to users_behavior (1).csv
['.config', 'users_behavior (1).csv', 'users_behavior.csv', 'drive', 'sample_data']


In [5]:
df = pd.read_csv('users_behavior.csv')


# This code allows us to display all the columns
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

print(df.shape)
print(df.head(5))
print(df.info())
print(df.describe())

(3214, 5)
   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None
             calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   757

This datasheet has 3214 rows, 5 coloumns. There is a complete non-null data. The columns data sets are:


*   `calls` : float64
*   `minutes`: float64
*   `messages`: float64
*   `mb_used`: float64
*   `is ultra`: int64




In [4]:
# Split the source data into a training set, a validation set, and a test set

features = df.drop(['last_price'], axis=1)
target = df['last_price']

# Split the data
features_train, features_test, target_train, target_test =  train_test_split(features, target, test_size=0.2, random_state=42)